c/core.go

```
@@ -0,0 +1,173 @@
+//
+// - start: The main function that is going to be executing in this child
+// goroutine
+//
+// The start function will receive a context.Context record that must be used to
+// receive a stop signal from the parent supervisor. Depending on the Shutdown
+// settings, if the start function does not respect the given context, the
+// supervision system will either block forever or leak goroutines after a
+// timeout has been reached.
+func New(name string, start func(context.Context) error, opts ...Opt) Spec {
+	return New1(name, func(ctx context.Context, notifyChildStart func()) error {
+		notifyChildStart()
+		return start(ctx)
+	}, opts...)
+}
+
+// New1 is similar to New, with the difference that the start function receives
+// a `notifyStart` callback to the child routine, this callback allows the child
+// goroutine to signal when it has officially started. This is essential when
+// you want to guarantee some bootstrap on thread initialization.
+func New1(
+	name string,
+	start func(context.Context, func()) error,
+	opts ...Opt,
+) Spec {
+	spec := Spec{}
+
+	if name == "" {
+		panic("Child cannot have empty name")
+	}
+	spec.name = name
+
+	if start == nil {
+		panic("Child cannot have empty start function")
+	}
+
+	// apply options
+	for _, optFn := range opts {
+		optFn(&spec)
+	}
+	spec.start = start
+
+	// return spec
+	return spec
+}
+
+// Name returns the specified name for a Child Spec
+func (cs Spec) Name() string {
+	return cs.name
+}
+
+func waitTimeout(
+	terminateCh chan struct{},
+) func(Shutdown) error {
+	return func(shutdown Shutdown) error {
+		switch shutdown.tag {
+		case infinityT:
+			// We wait forever for the result
+			<-terminateCh
+			return nil
+		case timeoutT:
+			// we wait until some duration
+			select {
+			case <-terminateCh:
+				return nil
+			case <-time.After(shutdown.duration):
+				return errors.New("Child shutdown timeout")
+			}
+		default:
+			panic("Invalid shutdown value received")
```

Add comment explaining rationale of panic instead of returning error